File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/BpaPdcStream/DataFrame.vb
'*******************************************************************************************************
'  DataPacket.vb - PDCstream data packet
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  11/12/2004 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports System.Buffer
Imports System.Text
Imports Tva.DateTime.Common
Imports Tva.Math.Common
Imports Tva.Phasors.Common
Imports Tva.Phasors.BpaPdcStream.Common

Namespace BpaPdcStream

    ' This is essentially a "row" of PMU data at a given timestamp
    <CLSCompliant(False), Serializable()> _
    Public Class DataFrame

        Inherits DataFrameBase
        Implements ICommonFrameHeader

        Private m_packetNumber As Byte
        Private m_sampleNumber As Int16
        Private m_legacyLabels As String()

        Public Sub New()

            MyBase.New(New DataCellCollection)
            m_packetNumber = 1

        End Sub

        Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

            MyBase.New(info, context)

            ' Deserialize data frame
            m_packetNumber = info.GetByte("packetNumber")
            m_sampleNumber = info.GetInt16("sampleNumber")

        End Sub

        Public Sub New(ByVal sampleNumber As Int16)

            MyClass.New()
            m_sampleNumber = sampleNumber

        End Sub

        ' If you are going to create multiple data packets, you can use this constructor
        ' Note that this only starts becoming necessary if you start hitting data size
        ' limits imposed by the nature of the transport protocol...
        Public Sub New(ByVal packetNumber As Byte, ByVal sampleNumber As Int16)

            MyClass.New(sampleNumber)
            Me.PacketNumber = packetNumber

        End Sub

        Public Sub New(ByVal parsedFrameHeader As ICommonFrameHeader, ByVal configurationFrame As IConfigurationFrame, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

            MyBase.New(New DataFrameParsingState(New DataCellCollection, parsedFrameHeader.FrameLength, configurationFrame, _
                AddressOf BpaPdcStream.DataCell.CreateNewDataCell), binaryImage, startIndex)

            CommonFrameHeader.Clone(parsedFrameHeader, Me)

        End Sub

        Public Sub New(ByVal dataFrame As IDataFrame)

            MyBase.New(dataFrame)

        End Sub

        Public Overrides ReadOnly Property DerivedType() As System.Type
            Get
                Return Me.GetType()
            End Get
        End Property

        Public Shadows ReadOnly Property Cells() As DataCellCollection
            Get
                Return MyBase.Cells
            End Get
        End Property

        Public Shadows Property ConfigurationFrame() As ConfigurationFrame
            Get
                Return MyBase.ConfigurationFrame
            End Get
            Set(ByVal value As ConfigurationFrame)
                MyBase.ConfigurationFrame = value
            End Set
        End Property

        Public Property PacketNumber() As Byte Implements ICommonFrameHeader.PacketNumber
            Get
                Return m_packetNumber
            End Get
            Set(ByVal value As Byte)
                m_packetNumber = value
            End Set
        End Property

        Public Property SampleNumber() As Int16 Implements ICommonFrameHeader.SampleNumber
            Get
                Return m_sampleNumber
            End Get
            Set(ByVal value As Int16)
                m_sampleNumber = value
            End Set
        End Property

        Public ReadOnly Property FrameType() As FrameType Implements ICommonFrameHeader.FrameType
            Get
                Return BpaPdcStream.FrameType.DataFrame
            End Get
        End Property

        Protected Overrides ReadOnly Property FundamentalFrameType() As FundamentalFrameType Implements ICommonFrameHeader.FundamentalFrameType
            Get
                Return MyBase.FundamentalFrameType
            End Get
        End Property

        Public Property WordCount() As Int16 Implements ICommonFrameHeader.WordCount
            Get
                Return MyBase.BinaryLength / 2
            End Get
            Set(ByVal value As Int16)
                MyBase.ParsedBinaryLength = value * 2
            End Set
        End Property

        Public ReadOnly Property FrameLength() As Short Implements ICommonFrameHeader.FrameLength
            Get
                Return MyBase.BinaryLength
            End Get
        End Property

        Public ReadOnly Property NtpTimeTag() As DateTime.NtpTimeTag
            Get
                Return New DateTime.NtpTimeTag(TicksToSeconds(Ticks))
            End Get
        End Property

        Public ReadOnly Property LegacyLabels() As String()
            Get
                Return m_legacyLabels
            End Get
        End Property

        <CLSCompliant(False)> _
        Protected Overrides Function CalculateChecksum(ByVal buffer() As Byte, ByVal offset As Int32, ByVal length As Int32) As UInt16

            ' PDCstream uses simple XOR checksum
            Return Xor16BitCheckSum(buffer, offset, length)

        End Function

        ' Oddly enough, check sum for frames in BPA PDC stream is little-endian
        Protected Overrides Sub AppendChecksum(ByVal buffer() As Byte, ByVal startIndex As Integer)

            EndianOrder.LittleEndian.CopyBytes(CalculateChecksum(buffer, 0, startIndex), buffer, startIndex)

        End Sub

        Protected Overrides Function ChecksumIsValid(ByVal buffer() As Byte, ByVal startIndex As Integer) As Boolean

            Dim sumLength As Int16 = BinaryLength - 2
            Return EndianOrder.LittleEndian.ToUInt16(buffer, startIndex + sumLength) = CalculateChecksum(buffer, startIndex, sumLength)

        End Function

        Protected Overrides ReadOnly Property HeaderLength() As UInt16
            Get
                If ConfigurationFrame.StreamType = StreamType.Legacy Then
                    Return 12 + ConfigurationFrame.Cells.Count * 8
                Else
                    Return 12
                End If
            End Get
        End Property

        Protected Overrides ReadOnly Property HeaderImage() As Byte()
            Get
                Dim buffer As Byte() = CreateArray(Of Byte)(HeaderLength)

                ' Common in common frame header portion of header image
                System.Buffer.BlockCopy(CommonFrameHeader.BinaryImage(Me), 0, buffer, 0, CommonFrameHeader.BinaryLength)

                If ConfigurationFrame.RevisionNumber = RevisionNumber.Revision0 Then
                    EndianOrder.BigEndian.CopyBytes(Convert.ToUInt32(NtpTimeTag.Value), buffer, 4)
                Else
                    EndianOrder.BigEndian.CopyBytes(Convert.ToUInt32(TimeTag.Value), buffer, 4)
                End If

                EndianOrder.BigEndian.CopyBytes(Convert.ToInt16(m_sampleNumber), buffer, 8)
                EndianOrder.BigEndian.CopyBytes(Convert.ToInt16(Cells.Count), buffer, 10)

                ' If producing a legacy format, include additional header
                If ConfigurationFrame.StreamType = StreamType.Legacy Then
                    Dim index As Integer = 12
                    Dim reservedBytes As Byte() = CreateArray(Of Byte)(2)
                    Dim offset As Int16

                    For x As Integer = 0 To Cells.Count - 1
                        With Cells(x)
                            CopyImage(Encoding.ASCII.GetBytes(.IDLabel), buffer, index, 4)
                            CopyImage(reservedBytes, buffer, index, 2)
                            EndianOrder.BigEndian.CopyBytes(offset, buffer, index)
                            index += 2
                            offset += .BinaryLength
                        End With
                    Next
                End If

                Return buffer
            End Get
        End Property

        Protected Overrides Sub ParseHeaderImage(ByVal state As IChannelParsingState, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

            Dim configurationFrame As BpaPdcStream.ConfigurationFrame = DirectCast(DirectCast(state, IDataFrameParsingState).ConfigurationFrame, BpaPdcStream.ConfigurationFrame)
            Dim dataCellCount As Int16

            ' Only need to parse what wan't already parsed in common frame header
            dataCellCount = EndianOrder.BigEndian.ToInt16(binaryImage, startIndex + 10)

            If dataCellCount <> configurationFrame.Cells.Count Then
                Throw New InvalidOperationException("Stream/Config File Mismatch: PMU count (" & dataCellCount & ") in stream does not match defined count in configuration file (" & configurationFrame.Cells.Count & ")")
            End If

            ' Note: because "HeaderLength" needs configuration frame and is called before associated configuration frame
            ' assignment normally occurs - we assign configuration frame in advance...
            Me.ConfigurationFrame = configurationFrame

            ' We'll at least retrieve legacy labels if defined (might be useful for debugging dynamic changes in data-stream)
            If configurationFrame.StreamType = StreamType.Legacy Then
                Dim index As Integer = 12

                For x As Integer = 0 To configurationFrame.Cells.Count - 1
                    m_legacyLabels = CreateArray(Of String)(configurationFrame.Cells.Count)
                    With configurationFrame.Cells(x)
                        m_legacyLabels(x) = Encoding.ASCII.GetString(binaryImage, index, 4)
                        index += 8
                    End With
                Next
            End If

        End Sub

        Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

            MyBase.GetObjectData(info, context)

            ' Serialize data frame
            info.AddValue("packetNumber", m_packetNumber)
            info.AddValue("sampleNumber", m_sampleNumber)

        End Sub

        Public Overrides ReadOnly Property Attributes() As System.Collections.Generic.Dictionary(Of String, String)
            Get
                Dim baseAttributes As Dictionary(Of String, String) = MyBase.Attributes

                baseAttributes.Add("Packet Number", m_packetNumber)
                baseAttributes.Add("Sample Number", m_sampleNumber)

                If m_legacyLabels IsNot Nothing Then
                    baseAttributes.Add("Legacy Label Count", m_legacyLabels.Length)

                    For x As Integer = 0 To m_legacyLabels.Length - 1
                        baseAttributes.Add("    Legacy Label " & x, m_legacyLabels(x))
                    Next
                End If

                Return baseAttributes
            End Get
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/BpaPdcStream/DataFrame.cs
//*******************************************************************************************************
//  DataFrame.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  11/12/2004 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.Serialization;
using System.Security.Permissions;
using System.Text;
using PCS.IO.Checksums;
using PCS.Parsing;

namespace PhasorProtocols.BpaPdcStream
{
    /// <summary>
    /// Represents the BPA PDCstream implementation of a <see cref="IDataFrame"/> that can be sent or received.
    /// </summary>
    [Serializable()]
    public class DataFrame : DataFrameBase, ISupportFrameImage<FrameType>
    {
        #region [ Members ]

        // Fields
        private CommonFrameHeader m_frameHeader;
        private ushort m_sampleNumber;
        private string[] m_legacyLabels;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="DataFrame"/>.
        /// </summary>
        /// <remarks>
        /// This constructor is used by <see cref="FrameImageParserBase{TTypeIdentifier,TOutputType}"/> to parse a BPA PDCstream data frame.
        /// </remarks>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public DataFrame()
            : base(new DataCellCollection(), 0, null)
        {
        }

        /// <summary>
        /// Creates a new <see cref="DataFrame"/> from specified parameters.
        /// </summary>
        /// <param name="timestamp">The exact timestamp, in <see cref="Ticks"/>, of the data represented by this <see cref="DataFrame"/>.</param>
        /// <param name="configurationFrame">The <see cref="ConfigurationFrame"/> associated with this <see cref="DataFrame"/>.</param>
        /// <param name="packetNumber">Packet number for this <see cref="DataFrame"/>.</param>
        /// <param name="sampleNumber">Sample number for thie <see cref="DataFrame"/>.</param>
        /// <remarks>
        /// This constructor is used by a consumer to generate a BPA PDCstream data frame.
        /// </remarks>
        public DataFrame(Ticks timestamp, ConfigurationFrame configurationFrame, byte packetNumber, ushort sampleNumber)
            : base(new DataCellCollection(), timestamp, configurationFrame)
        {
            PacketNumber = packetNumber;
            m_sampleNumber = sampleNumber;
        }

        /// <summary>
        /// Creates a new <see cref="DataFrame"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected DataFrame(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize data frame
            m_frameHeader = (CommonFrameHeader)info.GetValue("frameHeader", typeof(CommonFrameHeader));
            m_sampleNumber = info.GetUInt16("sampleNumber");
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets reference to the <see cref="DataCellCollection"/> for this <see cref="DataFrame"/>.
        /// </summary>
        public new DataCellCollection Cells
        {
            get
            {
                return base.Cells as DataCellCollection;
            }
        }

        /// <summary>
        /// Gets or sets <see cref="ConfigurationFrame"/> associated with this <see cref="DataFrame"/>.
        /// </summary>
        public new ConfigurationFrame ConfigurationFrame
        {
            get
            {
                return base.ConfigurationFrame as ConfigurationFrame;
            }
            set
            {
                base.ConfigurationFrame = value;
            }
        }

        /// <summary>
        /// Gets or sets the parsing state for the this <see cref="DataFrame"/>.
        /// </summary>
        public new DataFrameParsingState State
        {
            get
            {
                return base.State as DataFrameParsingState;
            }
            set
            {
                base.State = value;
            }
        }

        /// <summary>
        /// Gets the identifier that is used to identify the IEEE C37.118 frame.
        /// </summary>
        public FrameType TypeID
        {
            get
            {
                return BpaPdcStream.FrameType.DataFrame;
            }
        }

        /// <summary>
        /// Gets or sets current <see cref="CommonFrameHeader"/>.
        /// </summary>
        public CommonFrameHeader CommonHeader
        {
            get
            {
                // Make sure frame header exists - using base class timestamp to
                // prevent recursion (m_frameHeader doesn't exist yet)
                if (m_frameHeader == null)
                    m_frameHeader = new CommonFrameHeader(1);

                return m_frameHeader;
            }
            set
            {
                m_frameHeader = value;

                if (m_frameHeader != null)
                {
                    State = m_frameHeader.State as DataFrameParsingState;
                    Timestamp = m_frameHeader.RoughTimestamp;
                }
            }
        }

        // This interface implementation satisfies ISupportFrameImage<FrameType>.CommonHeader
        ICommonHeader<FrameType> ISupportFrameImage<FrameType>.CommonHeader
        {
            get
            {
                return CommonHeader;
            }
            set
            {
                CommonHeader = value as CommonFrameHeader;
            }
        }

        /// <summary>
        /// Gets or sets the packet number of this <see cref="DataFrame"/>.
        /// </summary>
        public byte PacketNumber
        {
            get
            {
                return CommonHeader.PacketNumber;
            }
            set
            {
                CommonHeader.PacketNumber = value;
            }
        }

        /// <summary>
        /// Gets or sets the sample number of this <see cref="DataFrame"/>.
        /// </summary>
        public ushort SampleNumber
        {
            get
            {
                return m_sampleNumber;
            }
            set
            {
                m_sampleNumber = value;
            }
        }

        /// <summary>
        /// Gets the timestamp of this <see cref="DataFrame"/> formatted as NTP.
        /// </summary>
        public NtpTimeTag NtpTimeTag
        {
            get
            {
                return new NtpTimeTag(Timestamp);
            }
        }

        /// <summary>
        /// Gets the legacy labels parsed from the <see cref="DataFrame"/>, if any.
        /// </summary>
        public string[] LegacyLabels
        {
            get
            {
                return m_legacyLabels;
            }
        }

        /// <summary>
        /// Gets the length of the <see cref="HeaderImage"/>.
        /// </summary>
        protected override int HeaderLength
        {
            get
            {
                if (ConfigurationFrame.StreamType == StreamType.Legacy)
                    return 12 + ConfigurationFrame.Cells.Count * 8;
                else
                    return 12;
            }
        }

        /// <summary>
        /// Gets the binary header image of the <see cref="DataFrame"/> object.
        /// </summary>
        protected override byte[] HeaderImage
        {
            get
            {
                byte[] buffer = new byte[HeaderLength];
                int index = 0;

                // Copy in common frame header portion of header image
                CommonHeader.BinaryImage.CopyImage(buffer, ref index, CommonFrameHeader.FixedLength);

                if (ConfigurationFrame.RevisionNumber == RevisionNumber.Revision0)
                    EndianOrder.BigEndian.CopyBytes((uint)Math.Truncate(NtpTimeTag.Value), buffer, 4);
                else
                    EndianOrder.BigEndian.CopyBytes((uint)Math.Truncate(TimeTag.Value), buffer, 4);

                EndianOrder.BigEndian.CopyBytes(m_sampleNumber, buffer, 8);
                EndianOrder.BigEndian.CopyBytes((ushort)Cells.Count, buffer, 10);

                index += 12;

                // If producing a legacy format, include additional header
                if (ConfigurationFrame.StreamType == StreamType.Legacy)
                {
                    byte[] reservedBytes = new byte[2];
                    ushort offset = 0;

                    for (int x = 0; x < Cells.Count; x++)
                    {
                        DataCell dataCell = Cells[x];
                        Encoding.ASCII.GetBytes(dataCell.IDLabel).CopyImage(buffer, ref index, 4);
                        reservedBytes.CopyImage(buffer, ref index, 2);
                        EndianOrder.BigEndian.CopyBytes(offset, buffer, index);
                        index += 2;
                        offset += (ushort)dataCell.BinaryLength;
                    }
                }

                return buffer;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="DataFrame"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                CommonHeader.AppendHeaderAttributes(baseAttributes);
                baseAttributes.Add("Sample Number", m_sampleNumber.ToString());

                if (m_legacyLabels != null)
                {
                    baseAttributes.Add("Legacy Label Count", m_legacyLabels.Length.ToString());

                    for (int x = 0; x < m_legacyLabels.Length; x++)
                    {
                        baseAttributes.Add("    Legacy Label " + x, m_legacyLabels[x]);
                    }
                }

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Parses the binary header image.
        /// </summary>
        /// <param name="binaryImage">Binary image to parse.</param>
        /// <param name="startIndex">Start index into <paramref name="binaryImage"/> to begin parsing.</param>
        /// <param name="length">Length of valid data within <paramref name="binaryImage"/>.</param>
        /// <returns>The length of the data that was parsed.</returns>
        protected override int ParseHeaderImage(byte[] binaryImage, int startIndex, int length)
        {
            IDataFrameParsingState state = State;
            int index = startIndex + CommonFrameHeader.FixedLength;

            // Only need to parse what wan't already parsed in common frame header
            ConfigurationFrame configurationFrame = state.ConfigurationFrame as ConfigurationFrame;

            // Parse frame timestamp
            uint secondOfCentury = EndianOrder.BigEndian.ToUInt32(binaryImage, index);
            m_sampleNumber = EndianOrder.BigEndian.ToUInt16(binaryImage, index + 4);
            index += 6;

            if (configurationFrame.RevisionNumber == RevisionNumber.Revision0)
                Timestamp = (new NtpTimeTag(secondOfCentury, 0)).ToDateTime().Ticks + (long)(m_sampleNumber * configurationFrame.TicksPerFrame);
            else
                Timestamp = (new UnixTimeTag(secondOfCentury)).ToDateTime().Ticks + (long)(m_sampleNumber * configurationFrame.TicksPerFrame);

            // Because in cases where PDCxchng is being used the data cell count will be smaller than the
            // configuration cell count - we save this count to calculate the offsets later
            state.CellCount = EndianOrder.BigEndian.ToUInt16(binaryImage, index);
            index += 2;

            if (state.CellCount > configurationFrame.Cells.Count)
                throw new InvalidOperationException("Stream/Config File Mismatch: PMU count (" + state.CellCount + ") in stream does not match defined count in configuration file (" + configurationFrame.Cells.Count + ")");
            
            // We'll at least retrieve legacy labels if defined (might be useful for debugging dynamic changes in data-stream)
            if (configurationFrame.StreamType == StreamType.Legacy)
            {
                m_legacyLabels = new string[state.CellCount];

                for (int x = 0; x < state.CellCount; x++)
                {
                    m_legacyLabels[x] = Encoding.ASCII.GetString(binaryImage, index, 4);
                    // We don't need offsets, so we skip them...
                    index += 8;
                }
            }

            return (index - startIndex);
        }

        /// <summary>
        /// Calculates checksum of given <paramref name="buffer"/>.
        /// </summary>
        /// <param name="buffer">Buffer image over which to calculate checksum.</param>
        /// <param name="offset">Start index into <paramref name="buffer"/> to calculate checksum.</param>
        /// <param name="length">Length of data within <paramref name="buffer"/> to calculate checksum.</param>
        /// <returns>Checksum over specified portion of <paramref name="buffer"/>.</returns>
        protected override ushort CalculateChecksum(byte[] buffer, int offset, int length)
        {
            // PDCstream uses a 16-bit XOR based check sum
            return buffer.Xor16CheckSum(offset, length);
        }

        /// <summary>
        /// Appends checksum onto <paramref name="buffer"/> starting at <paramref name="startIndex"/>.
        /// </summary>
        /// <param name="buffer">Buffer image on which to append checksum.</param>
        /// <param name="startIndex">Index into <paramref name="buffer"/> where checksum should be appended.</param>
        /// <remarks>
        /// We override default implementation since BPA PDCstream implements check sum for frames in little-endian.
        /// </remarks>
        protected override void AppendChecksum(byte[] buffer, int startIndex)
        {
            // Oddly enough, check sum for frames in BPA PDC stream is little-endian
            EndianOrder.LittleEndian.CopyBytes(CalculateChecksum(buffer, 0, startIndex), buffer, startIndex);
        }

        /// <summary>
        /// Determines if checksum in the <paramref name="buffer"/> is valid.
        /// </summary>
        /// <param name="buffer">Buffer image to validate.</param>
        /// <param name="startIndex">Start index into <paramref name="buffer"/> to perform checksum.</param>
        /// <returns>Flag that determines if checksum over <paramref name="buffer"/> is valid.</returns>
        /// <remarks>
        /// We override default implementation since BPA PDCstream implements check sum for frames in little-endian.
        /// </remarks>
        protected override bool ChecksumIsValid(byte[] buffer, int startIndex)
        {
            int sumLength = BinaryLength - 2;
            return EndianOrder.LittleEndian.ToUInt16(buffer, startIndex + sumLength) == CalculateChecksum(buffer, startIndex, sumLength);
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public override void GetObjectData(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize data frame
            info.AddValue("frameHeader", m_frameHeader, typeof(CommonFrameHeader));
            info.AddValue("sampleNumber", m_sampleNumber);
        }

        #endregion
    }
}